### PR TITLE
Close test coverage gaps to meet 80% threshold

### LINKED ISSUE #180: test-coverage
# Test Coverage

> Close coverage gaps in under-tested modules to bring all source files above the 80% line-coverage threshold and strengthen overall project reliability.

## Context

Dispatch is a TypeScript CLI tool for AI agent orchestration. It uses Vitest as its test runner with V8 coverage provider, tsup for bundling, and targets ES2022. The project enforces an 80% line-coverage threshold in `vitest.config.ts`.

The test suite lives in `src/tests/` with 43 test files totalling 1204 tests. Tests follow a consistent pattern: hoisted mocks via `vi.hoisted()`, module-level `vi.mock()` for external SDKs and internal helpers (especially `../helpers/logger.js`), `beforeEach` mock resets, and shared fixtures from `src/tests/fixtures.ts` (which provides `createMockProvider`, `createMockDatasource`, `createMockTask`, `createMockIssueDetails`).

The current coverage report (all files at 90.37% overall lines) reveals five modules significantly below or near the threshold:

| Module | Lines | Key gaps |
|--------|-------|----------|
| `src/agents/commit.ts` | 6.66% | `generate()` method (lines 75–131), `buildCommitPrompt()`, `parseCommitResponse()` |
| `src/providers/codex.ts` | 0% | Entire module untested — `boot()`, `createSession()`, `prompt()`, `cleanup()`, `listModels()` |
| `src/providers/detect.ts` | 33.33% | `checkProviderInstalled()` function body (lines 32–36) |
| `src/providers/opencode.ts` | 78.89% | `listModels()` (lines 39–64), model override parsing (lines 106–114) |
| `src/helpers/gitignore.ts` | 84.61% | Bare-entry dedup path (line 34), write-error warning path (line 40) |

The `@openai/codex` package has no library entry-point, so the project uses a manual mock at `src/__mocks__/@openai/codex.ts` (exporting a stub `AgentLoop` class) plus a Vite alias in `vitest.config.ts` to resolve it during tests.

## Why

The project enforces an 80% line-coverage threshold. Two provider modules (`codex.ts` at 0%, `detect.ts` at 33%) and the commit agent (`commit.ts` at 6.66%) are drastically below this bar. The `opencode.ts` provider sits at 78.89% — just under the threshold — due to an untested `listModels()` export. Closing these gaps ensures the coverage gate passes reliably, prevents regressions in critical AI provider and agent code paths, and brings the test suite in line with the project's own quality standards.

## Approach

Follow the established test authoring patterns already used by `claude.test.ts`, `copilot.test.ts`, and `opencode.test.ts`:

1. **Mock external SDKs at the module level** using `vi.hoisted()` + `vi.mock()`. Each provider test mocks its SDK (`@openai/codex`, `@github/copilot-sdk`, `@opencode-ai/sdk`) and `../helpers/logger.js`. The commit agent test should mock `node:fs/promises`, `node:crypto`, and the provider instance.

2. **Use the fixtures module** (`src/tests/fixtures.ts`) for shared mock factories (`createMockProvider`, `createMockIssueDetails`, `createMockTask`) where applicable, especially for the commit agent tests.

3. **Test the full ProviderInstance contract** for each provider: `boot()`, `createSession()`, `prompt()`, `cleanup()`, and `listModels()`. This mirrors the structure of the existing copilot and claude test files.

4. **Test exported utility functions directly** — `buildCommitPrompt()` and `parseCommitResponse()` from `commit.ts` are already exported and can be unit-tested with pure input/output assertions without any mocking.

5. **Mock `node:child_process`** for `detect.ts` tests, following the same pattern used in `prereqs.test.ts`.

6. **Extend existing test files** where a file already exists but has insufficient coverage (commit-agent, opencode, gitignore), and create new test files only for modules with zero test coverage (codex, detect).

## Integration Points

- **Vitest configuration** (`vitest.config.ts`): Coverage uses `v8` provider, reports `text` and `json`, includes `src/**/*.ts`, excludes `src/tests/**`, `src/**/interface.ts`, and `src/**/index.ts`. The `@openai/codex` alias points to `src/__mocks__/@openai/codex.ts`.
- **Mock stub** (`src/__mocks__/@openai/codex.ts`): Provides `AgentLoop` class with `constructor`, `run()`, and `terminate()` methods. The codex provider test should use `vi.mock()` to override this stub with test-specific behavior, following the same hoisted-mock pattern.
- **ProviderInstance interface** (`src/providers/interface.ts`): All provider tests validate conformance to `name`, `model`, `createSession()`, `prompt()`, `cleanup()`.
- **Agent interface** (`src/agents/interface.ts`): The commit agent extends `Agent` with a `generate()` method. Tests should validate both the base `Agent` contract and the `CommitAgent`-specific behavior.
- **Shared fixtures** (`src/tests/fixtures.ts`): `createMockProvider()`, `createMockTask()`, `createMockIssueDetails()` should be reused in commit agent tests.
- **Logger mock** (`../helpers/logger.js`): Every test file mocks the logger with `vi.fn()` stubs for `debug`, `warn`, `error`, `formatErrorChain`, `extractMessage`.
- **Test naming convention**: Files are named `<module>.test.ts` in `src/tests/`. Describe blocks mirror the module's exported functions or interface methods.

## Tasks

- [x] (P) Add comprehensive tests for the commit agent `generate()` method, `buildCommitPrompt()`, and `parseCommitResponse()` in `src/tests/commit-agent.test.ts`. Cover: successful generation with valid structured response, empty provider response, unparseable response (no commit message or PR title), provider error during `generate()`, prompt building with various issue/task combinations (with/without body, labels, completed/failed tasks, diff truncation), and response parsing edge cases (missing sections, extra whitespace, partial matches). Use `createMockProvider`, `createMockTask`, and `createMockIssueDetails` from fixtures. Mock `node:fs/promises` (`mkdir`, `writeFile`) and `node:crypto` (`randomUUID`). Commit with message: `test: add comprehensive commit agent coverage`.

- [x] (P) Create `src/tests/codex.test.ts` with full test coverage for the Codex provider. Follow the structure of `claude.test.ts` and `copilot.test.ts`. Mock `@openai/codex` via `vi.mock()` with a hoisted mock `AgentLoop` class. Test `listModels()` returns the hardcoded model list. Test `boot()` with default and custom model options. Test `createSession()` success and failure paths. Test `prompt()` with: successful message extraction from `AgentLoop.run()` items, session-not-found error, empty response, and `run()` rejection. Test `cleanup()` terminates all sessions and clears them, handles terminate errors gracefully, and is safe to call with no sessions. Commit with message: `test: add codex provider test coverage`.

- [x] (P) Create `src/tests/detect.test.ts` with tests for `checkProviderInstalled()` and the `PROVIDER_BINARIES` constant from `src/providers/detect.ts`. Mock `node:child_process` and `node:util` following the pattern in `prereqs.test.ts`. Test that it returns `true` when the binary executes successfully, returns `false` when execution fails, and validate each provider name maps to the expected binary. Commit with message: `test: add provider detection test coverage`.

- [x] (P) Extend `src/tests/opencode.test.ts` to cover the `listModels()` function and model override parsing in `boot()`. Add a new `describe("listModels", ...)` block testing: connection to existing server via URL, spawning a local server, filtering providers by source, returning empty array when data is null, cleanup of local server in finally block, and error propagation when createOpencode fails. Add tests for model override parsing: valid `provider/model` format sets `modelOverride`, invalid format (no slash) is ignored with a debug log. Commit with message: `test: improve opencode provider coverage`.

- [x] (P) Extend `src/tests/gitignore.test.ts` (or the existing test file that covers `ensureGitignoreEntry`) to add tests for the bare-entry deduplication path (entry without trailing slash already present) and the write-error warning path (mock `writeFile` to reject and verify `log.warn` is called). These correspond to the uncovered lines 34 and 40 in `src/helpers/gitignore.ts`. Commit with message: `test: improve gitignore helper coverage`.

- [x] (I) Run the full test suite with coverage (`npm run test:coverage`) to verify all tests pass and the 80% line-coverage threshold is met across all source files. Confirm no regressions in existing tests.

## References

- Vitest documentation: https://vitest.dev/guide/
- Vitest mocking guide: https://vitest.dev/guide/mocking
- V8 coverage provider: https://vitest.dev/guide/coverage
- Conventional Commits specification: https://www.conventionalcommits.org/
- Project coverage config: `vitest.config.ts` (thresholds, includes, excludes)

### PR DESCRIPTION
**Pull request by PatrickRuddiman**
## Summary

Closes #180

Adds comprehensive test coverage for several under-tested modules to bring all source files above the 80% line-coverage threshold enforced in `vitest.config.ts`.

## Changes

- **Commit agent** (`src/tests/commit-agent.test.ts`): Tests for `generate()`, `buildCommitPrompt()`, and `parseCommitResponse()` covering successful generation, empty/unparseable responses, provider errors, prompt building with various issue/task combinations, and response parsing edge cases.
- **Codex provider** (`src/tests/codex.test.ts`): Full coverage for `listModels()`, `boot()`, `createSession()`, `prompt()`, and `cleanup()` including success, failure, and edge-case paths.
- **Provider detection** (`src/tests/detect.test.ts`): Tests for `checkProviderInstalled()` and `PROVIDER_BINARIES` validation.
- **Opencode provider** (`src/tests/opencode.test.ts`): Extended with `listModels()` tests (server connection, spawning, filtering, cleanup, error propagation) and model override parsing tests.
- **Gitignore helper** (`src/tests/gitignore.test.ts`): Added tests for bare-entry deduplication and write-error warning paths.

## Verification

Full test suite passes with coverage (`npm run test:coverage`) and the 80% line-coverage threshold is met across all source files.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,308 +1,266 @@
...
+| Key | Description |
+|-----|-------------|
+| `provider` | AI backend: `opencode` (default), `copilot`, `claude`, or `codex` |
+| `source` | Issue tracker: `github`, `azdevops`, or `md` |
+| `concurrency` | Max parallel task dispatches |
+| `planTimeout` | Planning timeout in minutes (default: 10) |
+| `planRetries` | Retry attempts after planning timeout (default: 1) |
+| `org` | Azure DevOps organization URL |
+| `project` | Azure DevOps project name |
+| `serverUrl` | URL of a running provider server |
```
[the remark this answers is not in the document]
Comment (PatrickRuddiman) [reply README.md]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/PatrickRuddiman/Dispatch/pull/185#discussion_r2880442149)
